Hazel/src/Hazel/Window.h

```diff
@@ -35,13 +33,25 @@ namespace Hazel {
 		virtual unsigned int GetHeight() const = 0;
 
 		// Window attributes
-		virtual void SetEventCallback(const EventCallbackFn& callback) = 0;
+		virtual void SetEventCallback(void (*callback)(Event& e)) { m_Data.EventCallback = callback; };
 		virtual void SetVSync(bool enabled) = 0;
 		virtual bool IsVSync() const = 0;
 
 		virtual void* GetNativeWindow() const = 0;
 
 		static Window* Create(const WindowProps& props = WindowProps());
+
+	protected:
+		struct WindowData
```

Comment:
Why move this structure into Window? Seems like if thats useful it should be a separate PR

Reply:
I initially moved it since I was trying to use templates and those dont work with virtual functions. I can move it back.

Reply:
This should be in `WindowsWindow`, because different graphics APIs might want to store/handle this data in different ways and I think `Window` is meant to be a pure virtual interface so it shouldn't store any data.

Reply:
Since it is still here, I just want to clear it up: don't change things that are outside the scope of the PR.

As you mension, you did it to try something, failed, tried something else, it worked, congratz. The cleaning up part of a PR should include to remove unnecessary changes, so this should move back to WindowsWindow as Puddlestomper said.